### DISCUSSION TITLE
Book 3 - bug in final code? black image when using lambert material for nearest object to camera in mixed density pdf approach

### DISCUSSION THREAD

**Discussion by 1vx-437312114** [RayTracing/raytracing.github.io#1240] [Debugging]
This is the result of the final code of book 3 (_src/TheRestOfYourLife/main.cc_ ):

![theRestOfYourLife_400](https://github.com/RayTracing/raytracing.github.io/assets/23039163/ae03394e-ed38-4f6b-9feb-d2f5e3f59dea)

When i change the material of the glass sphere  in the `cornell_box()` - function into a lambert diffuse material i get a nearly black image with basically only the light source visible when i render the image:

![theRestOfYourLife](https://github.com/RayTracing/raytracing.github.io/assets/23039163/7b682b8b-3a12-435d-acad-7ef76d67b7e7)

This is what it should look like (rendered for comparison in the brute force path tracer without the mixed density pdf approach):

![cornell_box](https://github.com/RayTracing/raytracing.github.io/assets/23039163/52077f02-cd73-48a5-9eeb-ba24586957d9)

I tried a few things e.g. replacing the sphere with a box or changing materials of the box inside the scene to lambert diffuse. My results so far are that whenever the object which is nearest to the camera is lambert diffuse the rendering breaks and i get the nearly black image from above as the result.

I came to this after i tried implementing the mixed density pdf approach from book 3 in my own path tracer where i tried to render the original cornell box (in which everything has a diffuse material) and i got the same bad results which got me trying to test the original code.

Here are the "changes" (just one line) in the `cornell_box()` - function of the _main.cc_ file inside the folder  _src/TheRestOfYourLife/_ for changing the glas sphere in the front to a blue colored diffuse sphere:

   
```
 hittable_list cornell_box() {
    ...
    shared_ptr<material> blue = make_shared<lambertian>(color(0.0, 0.0, 0.75));   // blue lambert diffuse
    //auto glass = make_shared<dielectric>(1.5);
    objects.add(make_shared<sphere>(point3(190,90,190), 90 , blue));  // changed from 'glass' to 'blue'

    return objects;
}

```

Has anyone any idea what is going wrong here or what i did wrong and how to fix this? Thanks in advance.

Here is the "modified" main.cc if anyone wants to test it:

[main_cc.txt](https://github.com/RayTracing/raytracing.github.io/files/12439090/main_cc.txt)

**Comment by davidfrogley** [comment]
I got a similar result (all black, except for the light) as soon as I updated my code to include the "Sampling Directions towards a Hittable" section.

**Comment by hollasch** [comment]
Spawned issue #1265.
